lib/Sema/TypeCheckAttr.cpp

```diff
@@ -4241,11 +4241,15 @@ void AttributeChecker::visitNoDerivativeAttr(NoDerivativeAttr *attr) {
         diag::noderivative_only_on_differentiable_struct_or_class_fields);
     return;
   }
-  if (!conformsToDifferentiable(
-          nominal->getDeclaredInterfaceType(),
-          nominal->getDeclContext())) {
-    diagnoseAndRemoveAttr(attr,
-        diag::noderivative_only_on_differentiable_struct_or_class_fields);
+  // Find any `Differentiable` conformance for the nominal type. If no such
```

Comment:
This "nonlocal" check seems like it'll increase the risk of cross-file, cross-module, or circular typechecking bugs, even though there might not be any now. Could we simplify things by always allowing `@noDerivative` even if the struct/class is not `Differentiable`?

Reply:
> Could we simplify things by always allowing `@noDerivative` even if the struct/class is not `Differentiable`?

Thanks for the observation! I'd lean towards more conservative type-checking for `@noDerivative` (and attributes in general) instead of allowing it on all stored properties. How about revisiting this decision if more issues arise?

[Cross-file `Differentiable` derived conformances are disallowed](https://github.com/apple/swift/blob/24d48cab995cdbbf8df3f1ad5c77658c5e91ed1b/lib/Sema/DerivedConformances.cpp#L607), but this is a bit orthogonal to `@noDerivative` type-checking.

Reply:
Looking through `Attr.def`, I'm not sure there's precedent for domain-specific attributes like `@noDerivative`. The `CodingKeys` enum for `Encodable`/`Decodable` derived conformances is probably the closest semantic parallel.

`@IBAction` is also domain-specific and type-checking doesn't have requirements on the contextual type - I'm not sure it's an apples-to-apples comparison though.
```
class X {
  @IBAction // okay
  func foo(_ x: Float) {}
}
```

Reply:
Sounds good.